otherlibs/win32unix/socket.c

```diff
@@ -24,17 +25,54 @@ int socket_type_table[] = {
   SOCK_STREAM, SOCK_DGRAM, SOCK_RAW, SOCK_SEQPACKET
 };
 
-CAMLprim value unix_socket(value cloexec, value domain, value type, value proto)
+SOCKET win32_socket(int domain, int type, int protocol,
+                    LPWSAPROTOCOL_INFO info,
+                    BOOL inherit)
 {
   SOCKET s;
+  DWORD flags = WSA_FLAG_OVERLAPPED;
 
-  s = socket(socket_domain_table[Int_val(domain)],
-                   socket_type_table[Int_val(type)],
-                   Int_val(proto));
+#ifndef WSA_FLAG_NO_HANDLE_INHERIT
+#define WSA_FLAG_NO_HANDLE_INHERIT 0x80
+#endif
+
+  if (! inherit)
+    flags |= WSA_FLAG_NO_HANDLE_INHERIT;
+
+  s = WSASocket(domain, type, protocol, info, 0, flags);
   if (s == INVALID_SOCKET) {
-    win32_maperr(WSAGetLastError());
-    uerror("socket", Nothing);
+    if (! inherit && WSAGetLastError() == WSAEINVAL) {
+      /* WSASocket probably doesn't suport WSA_FLAG_NO_HANDLE_INHERIT,
+       * retry without. */
+      flags &= ~(DWORD)WSA_FLAG_NO_HANDLE_INHERIT;
```

Comment:
Is there an issue with the simpler
```suggestion
      flags &= ~WSA_FLAG_NO_HANDLE_INHERIT;
```
?

Reply:
You're right, it should work. `DWORD flags` is 4 bytes long, the constant is an int, so that shouldn't be a problem. I've taken up the habit of casting constants to the type of the result when fiddling with bits to avoid integer promotion, signedness, and size issues.